addons/web_editor/static/src/js/editor/snippets.editor.js

```diff
@@ -1723,25 +1783,61 @@ var SnippetsMenu = Widget.extend({
     _makeSnippetDraggable: function ($snippets) {
         var self = this;
         var $tumb = $snippets.find('.oe_snippet_thumbnail_img:first');
-        var left = $tumb.outerWidth() / 2;
-        var top = $tumb.outerHeight() / 2;
         var $toInsert, dropped, $snippet;
 
+        this._draggableCursorAt = {left: $tumb.outerWidth() / 2, top: $tumb.outerHeight() / 2};
+
         $snippets.draggable({
+            appendTo: this.$body,
+            cursor: 'move',
+            cursorAt: this._draggableCursorAt,
+            distance: 30,
+            drag: function (ev, ui) {
+                const draggableOffset = this._draggableCursorAt.top + ui.offset.top;
+                const scrollOffsetThreshold = 100;
+                const scrollStep = 10;
+                const $scrollTarget = $('html, body');
+                const scrollTimer = 10;
+                const visibleBottomOffset = $scrollTarget.scrollTop() +
+                                            $scrollTarget.get(0).clientHeight -
+                                            draggableOffset;
+                const visibleTopOffset = draggableOffset - $scrollTarget.scrollTop();
+
+                if (visibleTopOffset <= scrollOffsetThreshold + this._mainNavBarHeight) {
+                    this._clearAutoScrollInterval();
+                    this._autoScrollHandler = setInterval(function () {
+                            $scrollTarget.scrollTop($scrollTarget.scrollTop() -
+                                                    scrollStep *
+                                                    (
+                                                        1 - visibleTopOffset /
+                                                        (scrollOffsetThreshold + this._mainNavBarHeight)
+                                                    ));
+                        }.bind(this),
+                        scrollTimer
+                    );
+                } else if (visibleBottomOffset <= scrollOffsetThreshold) {
+                    this._clearAutoScrollInterval();
+                    this._autoScrollHandler = setInterval(function () {
+                            $scrollTarget.scrollTop($scrollTarget.scrollTop() +
+                                                    scrollStep *
+                                                    (
+                                                        1 - visibleBottomOffset / scrollOffsetThreshold
+                                                    ));
+                        }.bind(this),
+                        scrollTimer
+                    );
+                } else {
+                    this._clearAutoScrollInterval();
+                }
+            }.bind(this),
```

Comment:
Need to share the code with the other drag. Seems to be the exact same.
I also think the code can be simplified, what about:
- Starting a `setInterval` on drag start with a function being a merge of the two you defined. In both case you do: `$scrollTarget.scrollTop($scrollTarget.scrollTop() + a * (1 - b))`. Only the value of a and b changes.
- Ending the interval on drap stop
- On drag event, just update the value of `a` and `b`.

Reply:
Well I ended up refactoring everything into a Mixin. I pushed and checked everything.